lib/streamlit/elements/chat.py

```diff
@@ -0,0 +1,121 @@
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.runtime.state import (
+    WidgetArgs,
+    WidgetCallback,
+    WidgetKwargs,
+    register_widget,
+)
+from streamlit.type_util import Key, SupportsStr, to_key
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+FORM_DOCS_INFO: Final = """
+For more information, refer to the
+[documentation for forms](https://docs.streamlit.io/library/api-reference/control-flow/st.form).
+"""
+
+
+@dataclass
+class ChatInputSerde:
+    value: SupportsStr
+
+    def deserialize(self, ui_value: Optional[str], widget_id: str = "") -> str | None:
+        return ui_value
+
+    def serialize(self, v: str | None) -> str | None:
+        return v
+
+
+class ChatMixin:
+    @gather_metrics("chat_input")
+    def chat_input(
+        self,
+        placeholder: str | None = None,
+        *,
+        key: Key | None = None,
+        max_chars: int | None = None,
+        disabled: bool = False,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+    ) -> str | None:
+        # We default to an empty string here and disallow user choice intentionally
+        default = ""
+        key = to_key(key)
+        check_callback_rules(self.dg, on_change)
+
+        # We omit this check for scripts running outside streamlit, because
+        # they will have no script_run_ctx.
+        if runtime.exists():
+            if is_in_form(self.dg):
+                # It doesn't make sense to create a chat_input inside a form.
+                # We throw an error to warn the user about this.
+                raise StreamlitAPIException(
+                    f"`st.chat_input()` can't be used in an `st.form()`. {FORM_DOCS_INFO}"
+                )
+            if self.dg._active_dg == self.dg._main_dg:
+                pass
+
+        chat_input_proto = ChatInputProto()
+        if placeholder is not None:
+            chat_input_proto.placeholder = str(placeholder)
+
+        if max_chars is not None:
+            chat_input_proto.max_chars = max_chars
+
+        chat_input_proto.default = default
+        chat_input_proto.position = ChatInputProto.Position.BOTTOM
+
+        ctx = get_script_run_ctx()
+        serde = ChatInputSerde("")
+        widget_state = register_widget(
+            "chat_input",
+            chat_input_proto,
+            user_key=key,
+            on_change_handler=on_change,
+            args=args,
+            kwargs=kwargs,
+            deserializer=serde.deserialize,
+            serializer=serde.serialize,
+            ctx=ctx,
+        )
+
+        chat_input_proto.disabled = disabled
+        if widget_state.value_changed and widget_state.value is not None:
+            chat_input_proto.value = widget_state.value
+            chat_input_proto.set_value = True
+
+        self.dg._enqueue("chat_input", chat_input_proto)
+        return widget_state.value
```

Comment:
I believe we need to do this check:

```python
return None if widget_state.value_changed else widget_state.value
```

Otherwise, the set session state mechanism won't work here correctly. This is caused by the string trigger mechanism working a bit differently from normal text input that has a synchronized value on front- and backend. For example:

```
if st.button("Select template prompt"):
   st.session_state["my_chat_input"] = PROMPT_TEMPLATE
  
prompt = st.chat_input(key="my_chat_input")
```

The expectation would be that clicking the button would fill in the template prompt into the chat input field. And it's now on the user to submit this value (via `enter` or submit button). By directly returning `widget_state.value` here, the PROMPT_TEMPLATE would be inserted into the chat input in the frontend but also directly returned. This would cause the user script to already react to the new prompt without having the user actually submitted it.